toolchain/check/testdata/class/derived_to_base.carbon

```diff
@@ -0,0 +1,193 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+
+base class A {
+  var a: i32;
+}
+
+base class B {
+  extend base: A;
+  var b: i32;
+}
+
+class C {
+  extend base: B;
+  var c: i32;
+}
+
+fn ConvertCToB(p: C*) -> B* { return p; }
+fn ConvertBToA(p: B*) -> A* { return p; }
+fn ConvertCToA(p: C*) -> A* { return p; }
+
+fn ConvertValue(c: C) {
+  let a: A = c;
+}
+
+fn ConvertRef(c: C*) -> A* {
+  return &(*c as A);
+}
+
+fn ConvertInit() {
+  let a: A = {.base = {.base = {.a = 1}, .b = 2}, .c = 3} as C;
+}
+
```

Comment:
I think the code works for it, but should there be some tests of non-class-type bases? e.g.:

```suggestion

class StructBase {
  extend base: {.x: i32};
}

fn ConvertStructBaseValue(v: StructBase) {
  let s: {.x: i32} = v;
}

fn ConvertStructBase(p: StructBase*) -> {.x: i32}* { return p; }
```

Reply:
Done. Added to `fail_base_bad_type.carbon` because we don't allow such inheritance at all at the moment, but it seems good to make sure that it at least doesn't crash.

Reply:
LGTM, for some reason I was thinking it was allowed.